### ISSUE TITLE
Mongo backend memory efficiency

### ISSUE THREAD

**Issue by slipcon** [rrd4j/rrd4j#159]
We have a long-running application that stores a large number of metrics in multiple RRD databases, backed by Mongo.   We update the data once a minute.    We have another part of the same application that generates graphs on request, which typically would be much less often.

We've been doing some memory profiling recently, and are seeing some inefficiencies that we'd like to address, but not sure if its possible given the API - I wanted to be sure we're using it correctly, and ask for any pointers or suggestions.  Specifically, we're seeing a large amount of temporary memory being allocated and then garbage collected.

1.  There doesn't appear to be a way to generate a graph from an already in-memory RRD database:

e.g. 

```
    graphDef.datasource( rrdName, rrdDb.getUri(), rrdName, ConsolFun.AVERAGE, rrdFactory );
```

when the graph is generated, it has to load the RRDDB from Mongo - this allocates a bunch of memory.

2.  Because the graph generation reads directly from Mongo, that means every time we want to store data and have it graphed,  we need to close the RrdDb - but then in order to store the next sample, we have to reopen it:

```
    // doesn't write to mongo until close - which means we can't graph data from the current runtime unless we close & reopen.
     try
      {
        rrdDb.close();
      }
      catch ( IOException e )
      {
        log.error( "unable to close RRDDB: " + rrdName + ": " + e.getMessage() );
      }
    }


    RrdDef def = new RrdDef( rrdFactory.getUri( rrdName ) );

    try
    {
      RrdDb db = RrdDb.getBuilder().setPath( def.getUri() ).build();
      return db;
    }
    catch ( IOException e )
    {
      if ( e instanceof FileNotFoundException )
      {
        return null;
      }

      log.error( "caught", e );
    }
```

In our profiling,  both the close (which writes the byte array to Mongo) and the open ( which reads the byte array) allocate a decent amount of temporary memory.  (this status collection thread is the second highest in terms of bytes allocated over time in this test)

Ideally, we'd be able to do two things differently:

a) Graph from an already loaded RrdDatabase object
b) Keep a RrdDatabase object open for further samples,  but flush it to storage.   Currently we close it after every sample, both so that we can graph it, and to ensure persistence.    It would probably be ok in our use case to persist to the Mongo database less frequently, or maybe even only on shutdown,  if we could graph from memory.

**Comment by fbacchella** [comment]
Did you true to use a pool ?

**Comment by slipcon** [comment]
I'm not sure I understand what the pool would give me here -  I'm not having issues synchronizing the two uses, I don't think.      Its just that continuously closing & reopening the RRDDB ends up allocating a lot of temporary memory that then needs to be garbage collected.

**Comment by fbacchella** [comment]
Pool keeps a reference to the rrd, it handle a cache and return the same rrd. So if you keep a reference, you will save memory and access.

**Comment by slipcon** [comment]
I'm looking at the pool more now, and still don't think it solves my issue:

In my main (data sampling) thread - if I add a data sample, and release the RrdDb to the pool, presumably my usage will drop to 0, and the RrdDb will be closed:  this writes the db to mongo (good)  but then I can no longer add another data sample without reopening it (which reads the whole thing from Mongo again)

In my graphing thread, if I attempt to draw and it happens to be open, it'll block and then re-use the existing/open RrdDb data (good) - but if there is no contention (normally true) then it'll have to open the RrdDb from mongo every time.  

In other words, I think it really only saves an open every once in a while - when I happen to get a request to draw while the RrdDb is open for a sample.   Otherwise the pool will still be constantly opening & closing the database.

**Comment by fbacchella** [comment]
Perhaps you can keep Rrd references in a cache using weak or soft references.
